File: Hastighet/src/Hastighet/Core/Log.cpp
#include "Log.hpp"

/*
RESET_COLOR "\033[0m"
RED_TEXT "\033[1;31m"
GREEN_TEXT "\033[1;32m"
YELLOW_TEXT "\033[1;33m"
BLUE_TEXT "\033[1;34m"
WHITE_TEXT "\033[1;37m"
*/

namespace Hastighet{

    std::string Log::getLocalTime(){
        auto now = std::chrono::system_clock::now();
        std::time_t current_time = std::chrono::system_clock::to_time_t(now);
        struct tm time_info;
        localtime_s(&time_info, &current_time);
        char time_buffer[26];
        std::strftime(time_buffer, sizeof(time_buffer), "%H:%M:%S", &time_info);
        return time_buffer;
    }

    void Log::trace(const std::string& message, const std::string& sender){
        std::cout 
        << "\033[1;34m" //blue
        << "[" << getLocalTime() << "] " << sender << ": " << "\"" << message << "\""
        << "\033[0m" << "\n";
    }

    void Log::info(const std::string& message, const std::string& sender){
        std::cout 
        << "\033[1;32m" //green 
        << "[" << getLocalTime() << "] " << sender << ": " << "\"" << message << "\""
        << "\033[0m" << "\n";
    }

    void Log::warn(const std::string& message, const std::string& sender){
        std::cout 
        << "\033[1;33m" //yellow
        << "[" << getLocalTime() << "] " << sender << ": " << "\"" << message << "\""
        << "\033[0m" << "\n";
    }

    void Log::error(const std::string& message, const std::string& sender){
        std::cout 
        << "\033[1;31m" //red
        << "[" << getLocalTime() << "] " << sender << ": " << "\"" << message << "\""
        << "\033[0m" << "\n";
    }

}

File: Hastighet/Core/Log.cs
using System;

namespace Hastighet.Core
{
    public static class Log
    {

#if DEBUG
        public static bool IsDebugEnabled = true;
#else
        public static bool IsDebugEnabled = false;
#endif

        private static void PrintWithFormating(string message, string sender)
        {
            Console.WriteLine($"[{DateTime.Now.ToString("T")}] {sender}: {message}");
            Console.ResetColor();
        }

        public static void Trace<T>(T message, string sender = "Hastighet")
        {
            if (IsDebugEnabled)
            {
                Console.ResetColor();
                PrintWithFormating(message.ToString(), sender);
            }
        }

        public static void Info<T>(T message, string sender = "Hastighet")
        {
            if (IsDebugEnabled)
            {
                Console.ForegroundColor = ConsoleColor.Green;
                PrintWithFormating(message.ToString(), sender);
            }
        }

        public static void Warn<T>(T message, string sender = "Hastighet")
        {
            if (IsDebugEnabled)
            {
                Console.ForegroundColor = ConsoleColor.Yellow;
                PrintWithFormating(message.ToString(), sender);
            }
        }

        public static void Error<T>(T message, string sender = "Hastighet")
        {
            if (IsDebugEnabled)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                if (message is Exception exception)
                {
                    PrintWithFormating(exception.Message, sender);
                }
                else
                {
                    PrintWithFormating(message.ToString(), sender);
                }
            }
        }

    }
}
